### PR TITLE
Jenkins-2 part of JENKINS-29649 + Fixed JENKINS-37623, JENKINS-38149

### PR DESCRIPTION
**Pull request by anenviousguest**
Hi @kinow 

Thanks for merging #13. Turns out there was missing part there, namely the fix for Jenkins 2 workflow jobs. Would appreciate if you could review it.

Also i put in here the fix for JENKINS-37623 , as the it and the change above modify same file.
# TL;DR

For Jenkins 2 workflow jobs, it might happen that JUnit and TAP publishers wrongly use TAP-only (or JUnit-only) results to draw their graphs. Added test case to detect that.

Also, when build history contained runs which failed before any TAP results were published, `NullPointerException` was occurring during the tooltip generation. The graph was still rendered, but the image map was missing.
# Why it happens (JENKINS-29649 part)

In JUnit's code inside `TestResultProjectAction`, they try to fetch last test result action of specific type before drawing a plot:
- https://github.com/jenkinsci/junit-plugin/blob/master/src/main/java/hudson/tasks/test/TestResultProjectAction.java#L109
- https://github.com/jenkinsci/junit-plugin/blob/master/src/main/java/hudson/tasks/test/TestResultProjectAction.java#L120

As one may see here: https://github.com/jenkinsci/junit-plugin/blob/master/src/main/java/hudson/tasks/test/TestResultProjectAction.java#L93 - the action they search for is supposed to be inherited from `AbstractTestResultAction`. 

And because `TapTestResultAction` is also derived from `AbstractTestResultAction`, it may (depending on the action order) be unexpectedly used to plot JUnit graphs, too. Vice versa might also be the case.
# This PR's content
1. The fix for Jenkins-2 part of JENKINS-29649 + test case which is red when fix is not applied. I basically decoupled `TapTestResultAction` from JUnit and migrated it and its surroundings to use generic `Run` instead of `AbstractProject`.
2. The fix for JENKINS-37623 + test case which is red when fix is not applied.
3. Because of the test cases above, I had to add test dependency and increase the minimum Jenkins version plugin is built against (still Jenkins 1.xxx though).
4. The new Jenkins parent seem to bring in `InjectedTest` which performs basic lint checks on *.jelly files. Had to add Jelly escaping directive to make this test happy.
5. Few IDE warnings fixed.

### CODE REVIEW

**Comment by kinow** [comment]
:+1: thanks @anenviousguest will need a few days, as I'm going through a few other plug-ins and pull requests to cut more releases. Feel free to update the pull request in the meantime.

**Comment by anenviousguest** [comment]
hi @kinow 
Have you by any chance had a time to look at this PR? Would very much appreciate if you could give your feedback or merge it.
Thanks!

**Comment by anenviousguest** [comment]
hi Bruno @kinow 
Let me kindly bump the thread here. Would that be reasonable to assume that you would have time to review this in a foreseeable future?
Thanks!

**Comment by kinow** [comment]
Hi @anenviousguest 

Ack :-) Promise that as soon as I'm done with https://github.com/jenkinsci/tap-plugin/tree/performance-fix for the performance issues.

Normally I make little progress, and proceed with other issues. But this time I really want to fix these performance issues. Once it's done, will continue merging the other PR's and working on issues. Yours will be the first PR that I will look into, and with RERO motto in mind, we might have it released shortly after the PR has been merged.

Sorry for the delay

File: pom.xml [hunk 1]
```diff
@@ -20,7 +20,7 @@
 	<parent>
 		<groupId>org.jenkins-ci.plugins</groupId>
 		<artifactId>plugin</artifactId>
-		<version>1.580.3</version><!-- which version of Jenkins is this plugin 
+		<version>1.637</version><!-- which version of Jenkins is this plugin
```
Comment (oleg-nenashev) [inline pom.xml]: 🐜 for selecting non-LTS version.
What was the reason for it?

File: src/main/java/org/tap4j/plugin/TapTestResultAction.java [hunk 2]
```diff
@@ -47,26 +50,31 @@
  * @since 0.1
  */
 public class TapTestResultAction
-        extends AbstractTestResultAction<AbstractTestResultAction<?>>
-        implements StaplerProxy, SimpleBuildStep.LastBuildAction {
+        implements StaplerProxy, SimpleBuildStep.LastBuildAction, HealthReportingAction, RunAction2 {
```
Comment (oleg-nenashev) [inline src/main/java/org/tap4j/plugin/TapTestResultAction.java]: Removal of AbstractTestResultAction inheritance smells like a compatibility issue, but seems it's reasonable in this case

File: src/main/java/org/tap4j/plugin/TapTestResultAction.java [hunk 3]
```diff
@@ -47,26 +50,31 @@
...
+
+    public transient Run<?,?> run;
+    @Deprecated
+    public transient AbstractBuild<?,?> owner;
```
Comment (oleg-nenashev) [inline src/main/java/org/tap4j/plugin/TapTestResultAction.java]: Well, the binary compatibility is broken in any case

File: src/main/java/org/tap4j/plugin/TapTestResultAction.java [hunk 4]
```diff
@@ -47,26 +50,31 @@
...
 
     private TapResult tapResult;
-    
+
+
     /**
-     * @param owner
+     * @param r
```
Comment (oleg-nenashev) [inline src/main/java/org/tap4j/plugin/TapTestResultAction.java]: 🐛 It changes param name to the invalid one.
Having of param declarations without description is also a bad practice. Javadoc linter in Java8 will grumble about that.

File: src/main/java/org/tap4j/plugin/TapTestResultAction.java [hunk 5]
```diff
@@ -160,88 +164,45 @@ public String getDisplayName() {
...
+        }
+        return new HealthReport(score, description);
     }
 
-    /* (non-Javadoc)
-     * @see hudson.tasks.test.AbstractTestResultAction#getTotalCount()
-     */
     @Override
-    @Exported(visibility = 2)
-    public int getTotalCount() {
```
Comment (oleg-nenashev) [inline src/main/java/org/tap4j/plugin/TapTestResultAction.java]: Binary compat as well

File: src/main/java/org/tap4j/plugin/model/TapTestResultResult.java [hunk 6]
```diff
@@ -279,31 +278,46 @@ public String toString() {
...
         return pw.toString();
     }
 
+    /**
+     * This is mostly a verbatim from {@link TestObject#getTestResultAction()}, with the only difference in a return
+     * type.
+     *
+     * @return associated TAP test result action object
+     */
+    private TapTestResultAction getTestResultActionDiverged() {
```
Comment (oleg-nenashev) [inline src/main/java/org/tap4j/plugin/model/TapTestResultResult.java]: Good practice: CheckForNull annotation

**Comment by oleg-nenashev** [comment]
The PR also addresses [JENKINS-38149](https://issues.jenkins-ci.org/browse/JENKINS-38149)

**Comment by anenviousguest** [comment]
@oleg-nenashev thanks for reviewing this. I'll try to accomodate changes where it deems reasonable.

File: src/main/java/org/tap4j/plugin/TapTestResultAction.java (hunk 2 above)
Comment (anenviousguest) [inline src/main/java/org/tap4j/plugin/TapTestResultAction.java]: @oleg-nenashev please see "Why it happens (JENKINS-29649 part)" above regarding `AbstractTestResultAction`

File: src/main/java/org/tap4j/plugin/TapTestResultAction.java (hunk 5 above)
Comment (anenviousguest) [inline src/main/java/org/tap4j/plugin/TapTestResultAction.java]: i guess it's a diff display issue. The lines highlighted were in the unused descendant class.
`getTotalCount` is still there in `TapTestResultAction`

File: src/main/java/org/tap4j/plugin/model/TapTestResultResult.java (hunk 6 above)
Comment (anenviousguest) [inline src/main/java/org/tap4j/plugin/model/TapTestResultResult.java]: addressed

File: src/main/java/org/tap4j/plugin/TapTestResultAction.java (hunk 2 above)
Comment (anenviousguest) [inline src/main/java/org/tap4j/plugin/TapTestResultAction.java]: was exactly causing the issue, please see "Why it happens (JENKINS-29649 part)" in the description above.

File: src/main/java/org/tap4j/plugin/TapTestResultAction.java (hunk 3 above)
Comment (anenviousguest) [inline src/main/java/org/tap4j/plugin/TapTestResultAction.java]: why? these have already been there, just inherited from `AbstractTestResultAction`. Now that it was removed from inheritance hierarchy, i placed these fields right here.

File: src/main/java/org/tap4j/plugin/TapTestResultAction.java (hunk 4 above)
Comment (anenviousguest) [inline src/main/java/org/tap4j/plugin/TapTestResultAction.java]: addressed

File: pom.xml (hunk 1 above)
Comment (anenviousguest) [inline pom.xml]: addressed.

**Comment by anenviousguest** [comment]
@kinow @oleg-nenashev does that look better now? Thanks.

File: src/main/java/org/tap4j/plugin/TapTestResultAction.java (hunk 2 above)
Comment (oleg-nenashev) [inline src/main/java/org/tap4j/plugin/TapTestResultAction.java]: Yes, I understand why this change is required. The problem is that...
* The class cannot be longer casted to AbstractTestResultAction
* Usage of non-implemented methods from AbstractTestResultAction will cause LinkageError. E.g. <code>findPreviousCorresponding()</code>

So I would vouch for keeping AbstractTestResultAction inheritance. If you update JUnit plugin dependency to the latest version (1.18), you may see it's interfaces have been adapted to make this class compatible with Pipeline Plugin

File: src/main/java/org/tap4j/plugin/TapTestResultAction.java (hunk 3 above)
Comment (oleg-nenashev) [inline src/main/java/org/tap4j/plugin/TapTestResultAction.java]: see above

**Comment by oleg-nenashev** [comment]
@anenviousguest Provided feedback. I think that the inheritance from <code>AbstractTestResultAction</code> should be retained. Upgrade to one of the latest JUnit plugin versions should help with it

File: src/main/java/org/tap4j/plugin/TapTestResultAction.java (hunk 2 above)
Comment (anenviousguest) [inline src/main/java/org/tap4j/plugin/TapTestResultAction.java]: hi @oleg-nenashev ,
I've just tested, and upgrading to junit-plugin 1.18 doesn't fix JENKINS-29649 in this plugin.

Everything described in the "why it happens..." at the very beginning of this PR  is still relevant to version 1.18.

That is, when a build contains multiple test result actions derived from `AbstractTestResultAction`, the return value of `TestResultProjectAction.getLastTestResultAction`, as seen [here](https://github.com/jenkinsci/junit-plugin/blob/master/src/main/java/hudson/tasks/test/TestResultProjectAction.java#L93) may depend on the order in which actions were saved, because it searches for actions of a base class.

Consequently, [here](https://github.com/jenkinsci/junit-plugin/blob/master/src/main/java/hudson/tasks/test/TestResultProjectAction.java#L120) and [here](https://github.com/jenkinsci/junit-plugin/blob/master/src/main/java/hudson/tasks/test/TestResultProjectAction.java#L109) wrong action might be used to draw the graph.

**Comment by anenviousguest** [comment]
@oleg-nenashev thanks! See my reply in the above thread.

**Comment by oleg-nenashev** [comment]
Well, we need to think about potential solutions. Now particular Jenkins features may stop working for TAP Plugin (e.g. aggregation of downstream test results in Freestyle jobs).

P.S: We've tested the change on a Pipeline-only instance, and the plugin is working as we need with this patch.

**Comment by kinow** [comment]
Will try to catch up with the review/comments this weekend :)

**Comment by anenviousguest** [comment]
Hi @oleg-nenashev 

I'm struggling to find any immediate elegant solution to work around [this check](https://github.com/jenkinsci/junit-plugin/blob/master/src/main/java/hudson/tasks/test/TestResultProjectAction.java#L93l) in junit-plugin.

I have no idea if JUnit plugin was meant to provide a base API for test-alike plugins, but if it was, it's probably better to think of fixing the check on its side, in a long term.

**Comment by oleg-nenashev** [comment]
Yes, I should have read the text carefully from the beginning.

### The JUnit Issue

> I have no idea if JUnit plugin was meant to provide a base API for test-alike plugins, but if it was, it's probably better to think of fixing the check on its side, in a long term. 

JUnit plugin has been decoupled from the Jenkins core at some point. So yes, it's actually a kind of API plugin for all test results then just a JUnit plugin.

> And because TapTestResultAction is also derived from AbstractTestResultAction, it may (depending on the action order) be unexpectedly used to plot JUnit graphs, too. Vice versa might also be the case.

It is a case.

### How to solve it?

Well, it should be solved on the JUnit plugin level. Just because of that: https://github.com/search?q=org%3Ajenkinsci+%22extends+AbstractTestResultAction%22&type=Code

I would just ignore this issue by now. I'll see if I can fix it on the JUnit plugin side by merging results in the plot (or by filtering them)

**Comment by anenviousguest** [comment]
hi @oleg-nenashev 

I guess i know what type of fix it might be on junit-plugin side. Should I just submit a PR there and if it's merged, revise this one?

Thanks.

**Comment by oleg-nenashev** [comment]
@anenviousguest Up to you. I think we can go forward with this PR even without waiting for JUnit stuff to be integrated and released

**Comment by alappe** [comment]
Any news here? A fix would be deeply appreciated…

**Comment by kinow** [comment]
Lurking in the discussion between @oleg-nenashev  and @anenviousguest :-) happy to review/merge whenever they confirm it's ready

**Comment by oleg-nenashev** [comment]
I think it's ready. The fix is required on the JUnit plugin. Merge of this PR does not increase the technical debt

**Comment by kinow** [comment]
Can we release after merging it @oleg-nenashev ? Or would it be necessary to wait for a merge and release in JUnit plugin first?

**Comment by oleg-nenashev** [comment]
no need in waiting for JUnit plugin IMHO. Just 🚢 🇮🇹  :)

**Comment by mkhan446** [comment]
Any updates on this pull request? I'm encountering JENKINS-38149 and would love to have it resolved... :)

**Comment by kinow** [comment]
Sorry @mkhan446 got a bit busy with active-choices-plugin, had to digress for a while, but also want to get this merged and released soon. Just give me a few more days and I will sort out and prepare a release to the experimental update center, so others can test it.

**Comment by anenviousguest** [comment]
@oleg-nenashev : thanks!
@kinow : i would very much appreciate if you could merge this, indeed :-) I've been running this fix over the past few months, and it seems to be working, just like @oleg-nenashev mentioned above, too.
Thanks guys for your cooperation!

**Comment by kinow** [comment]
Merged. Merge conflicts fixed manually, and all tests passed after that as well. Re-read the issue history, and the only concern is regarding previous behavior that may change after this change. Will push to experimental update center in a few minutes for now, and will play with the new plug-in before cutting a release, likely by the mid or end of December. Any feedback from the version in the experimental update center will definitely help! Thanks @anenviousguest and @oleg-nenashev !

**Comment by jglick** [comment]
@kinow please also amend https://github.com/jenkinsci/pipeline-plugin/pull/403 to indicate the status of the release (fix version).
